Attach sequence number in log output / collective mismatch error

o-hanna opened the pull request:
Fixes 1. in #72435 by attaching sequence number in log output / collective mismatch error. 

Now the log output should look like:
"
Detected mismatch between collectives on ranks. Rank 1 is running inconsistent collective: CollectiveFingerPrint(OpType=ALLREDUCE, TensorShape=[20, 10], TensorDtypes=Float, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt)). The sequence number is: 0
"

If the sequence number is not supported, it will attach:
"
ProcessGroup <backend_name> does not yet support sequence numbers
"


To run test:
python test/distributed/test_pg_wrapper.py

H-Huang reviewed (commented):
Thanks for creating the PR! Added a few comments

> Now the log output should look like: ...

Also the code/log for this has been updated so you should rebase the PR.

o-hanna commented on torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -90,6 +90,14 @@ struct CollectiveFingerPrint {
+            int SequenceNumberForGroup = pg->getSequenceNumberForGroup();
+            ss << ". The sequence number is: " << SequenceNumberForGroup;
+          } catch (c10::Error) {
+            ss << ". ProcessGroup " << pg->getBackendName()
```
Done.

o-hanna commented:
Updates: - Rebased the PR.
- If sequence number is not implemented we do not print anything.
- getSequenceNumberForGroup() is implemented in ProcessGroup base class but it throws an error by default.

Howard suggestions:
1. Change the API to return an std::optional<int> value and return std::nullopt to signify a nonexistent sequence number, but this would be BC breaking.
2. Change the default API to return a -1 to signify nonexistent sequence number (but a magic number to symbolize non-existence is usually frowned upon).
3. Update process group MPI to return a -1 (which is the only internal PG that does not support sequence number?).
4. Leave code as-is.

rohan-varma reviewed (commented):
Thanks for the PR and working on this! 

In addition to @H-Huang's comment, let's also add a unittest in `test_pg_wrapper.py` to ensure the sequence number is logged when this error is thrown.

mrshenli reviewed (changes requested):
I assumes changes to `third_party/kineto` are not intentional? Shall we get rid of those? Thanks.

o-hanna committed:
added sequence number

o-hanna committed:
added test to ensure the sequence number is logged

awgu commented on test/distributed/test_pg_wrapper.py:
```diff
@@ -33,6 +33,14 @@ def setUp(self):
+        seq_num_msg = "The sequence number is: "
+        seq_num_msg_index = err.find(seq_num_msg)
+        seq_num_msg_len = len(seq_num_msg)
+        self.assertTrue(seq_num_msg_index != -1, "Sequence number is missing.")
```
nit: 
```suggestion
        self.assertNotEqual(seq_num_msg_index, -1, "Sequence number is missing.")
```

awgu commented on test/distributed/test_pg_wrapper.py:
```diff
@@ -33,6 +33,14 @@ def setUp(self):
+        seq_num_msg_index = err.find(seq_num_msg)
+        seq_num_msg_len = len(seq_num_msg)
+        self.assertTrue(seq_num_msg_index != -1, "Sequence number is missing.")
+        if seq_num_msg_index != -1:
```
nit; I think this check is unneeded since if `seq_num_msg_index == -1`, then the previous assert will have already failed, in which case the test does not continue.

o-hanna replied:
Done. Thank you.

o-hanna committed:
added test to ensure the sequence number is logged

o-hanna committed:
Update test/distributed/test_pg_wrapper.py

Co-authored-by: Andrew Gu <31054793+awgu@users.noreply.github.com>

o-hanna commented:
Thank you for the comments.

Updates:
- Added test in test_pg_wrapper.py to ensure the sequence number is logged. The test runs successfully.
- Reverted changes in third_party/kineto.

rohan-varma reviewed (approved):
LGTM, pls get a stamp from @H-Huang as well to confirm he is okay with the try/catch approach

H-Huang reviewed (approved):
LGTM, thanks for updating!

mrshenli commented on torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:
```diff
@@ -90,6 +90,14 @@ struct CollectiveFingerPrint {
              << pg->getRank()
              << " is running inconsistent collective: " << *this;
+          // Attach sequence number in log output / collective mismatch error
+          try {
```
+1 to @H-Huang's suggestion

o-hanna replied:
Thank you for the comments. 
There is a virtual function getSequenceNumberForGroup() implemented in ProcessGroup base class, but it throws an error (“ProcessGroup .. does not yet support sequence numbers”); other classes override it. We left it as is to keep the non-existence of sequence number signified.
